control_panel_api/aws.py

```diff
@@ -120,3 +141,123 @@ def detach_policy_from_user(self, policy_arn, user_name):
+
+        if document:
+            self._load(document)
+
+    def revoke_access(self, bucket_arn):
+        self._readonly.discard(bucket_arn)
+        self._readwrite.discard(bucket_arn)
+
+    def grant_access(self, bucket_arn, readwrite=False):
+        self.revoke_access(bucket_arn)
+
+        if readwrite:
+            self._readwrite.add(bucket_arn)
+        else:
+            self._readonly.add(bucket_arn)
+
+    @property
+    def document(self):
+        statements = [
+            {
+                "Sid": "console",
+                "Effect": "Allow",
+                "Action": [
+                    "s3:GetBucketLocation",
+                    "s3:ListAllMyBuckets",
+                ],
+                "Resource": "arn:aws:s3:::*",
+            },
+        ]
+
+        all_buckets = self._readonly | self._readwrite
+        if all_buckets:
+            statements.append(self._list_statement)
+
+        if self._readonly:
+            statements.append(self._readonly_statement)
+
+        if self._readwrite:
+            statements.append(self._readwrite_statement)
+
+        return {
+            "Version": "2012-10-17",
+            "Statement": statements,
+        }
+
+    @property
+    def _list_statement(self):
+        all_buckets_arns = self._readonly | self._readwrite
+        return {
+            "Sid": "list",
+            "Action": [
+                "s3:ListBucket"
+            ],
+            "Effect": "Allow",
+            "Resource": list(all_buckets_arns),
+        }
+
+    @property
+    def _readonly_statement(self):
+        return {
+            "Sid": "readonly",
+            "Action": [
+                "s3:GetObject",
+                "s3:GetObjectAcl",
+                "s3:GetObjectVersion",
+            ],
+            "Effect": "Allow",
+            "Resource": self._with_star(self._readonly),
+        }
+
+    @property
+    def _readwrite_statement(self):
+        return {
+            "Sid": "readwrite",
+            "Action": [
+                "s3:GetObject",
+                "s3:GetObjectAcl",
+                "s3:GetObjectVersion",
+                "s3:DeleteObject",
+                "s3:DeleteObjectVersion",
+                "s3:PutObject",
+                "s3:PutObjectAcl",
+                "s3:RestoreObject",
+            ],
+            "Effect": "Allow",
+            "Resource": self._with_star(self._readwrite),
+        }
+
+    def _with_star(self, arns):
+        return [f'{arn}/*' for arn in arns]
+
+    def _without_star(self, resources):
+        result = []
+        for resource in resources:
+            without = resource
+            if resource.endswith('/*'):
+                without = resource[:-2]
+
+            result.append(without)
+
+        return result
+
+    def _load(self, document):
+        for statement in document['Statement']:
+            sid = statement['Sid']
+            if sid not in ['readonly', 'readwrite']:
+                continue
```

Comment:
Good point, not sure why I ended up the other way around :P 

Interesting here that you're using a tuple in your example, it actually makes sense to use them instead of lists as this "list" will not change and I also imagine tuples are faster.